bpf/aya-bpf/src/maps/perf/perf_event_array.rs

```diff
@@ -50,20 +57,54 @@ impl<T> PerfEventArray<T> {
         }
     }
 
-    pub fn output<C: BpfContext>(&self, ctx: &C, data: &T, flags: u32) {
-        self.output_at_index(ctx, BPF_F_CURRENT_CPU as u32, data, flags)
+    pub fn output_current_cpu<C: BpfContext>(&self, ctx: &C, data: &T) -> Result<(), i64> {
+        self.output(ctx, data, BPF_F_CURRENT_CPU)
+    }
+
+    pub fn output_at_index<C: BpfContext>(&self, ctx: &C, data: &T, index: u32) -> Result<(), i64> {
+        self.output(ctx, data, (index as u64) & BPF_F_INDEX_MASK)
     }
 
-    pub fn output_at_index<C: BpfContext>(&self, ctx: &C, index: u32, data: &T, flags: u32) {
-        let flags = (flags as u64) << 32 | index as u64;
+    fn output<C: BpfContext>(&self, ctx: &C, data: &T, flags: u64) -> Result<(), i64> {
         unsafe {
-            bpf_perf_event_output(
+            let ret = bpf_perf_event_output(
                 ctx.as_ptr(),
                 self.def.get() as *mut _,
                 flags,
                 data as *const _ as *mut _,
                 mem::size_of::<T>() as u64,
             );
+            if ret == 0 {
+                Ok(())
+            } else {
+                Err(ret)
+            }
+        }
+    }
+
+    pub fn read_current_cpu(&self) -> Result<bpf_perf_event_value, i64> {
+        self.read(BPF_F_CURRENT_CPU)
+    }
+
+    pub fn read_at_index(&self, index: u32) -> Result<bpf_perf_event_value, i64> {
+        self.read((index as u64) & BPF_F_INDEX_MASK)
+    }
+
+    fn read(&self, flags: u64) -> Result<bpf_perf_event_value, i64> {
+        let mut buf = MaybeUninit::<bpf_perf_event_value>::uninit();
+        unsafe {
+            // According to the Linux manual, `bpf_perf_event_read_value` is preferred over `bpf_perf_event_read`.
+            let ret = bpf_perf_event_read_value(
+                self.def.get() as *mut _,
+                flags,
+                buf.as_mut_ptr(),
+                mem::size_of::<bpf_perf_event_value>() as u32,
+            );
+            if ret == 0 {
+                Ok(buf.assume_init())
+            } else {
+                Err(ret)
```

Comment:
I think this is wrong and this should grab the errno.

```suggestion
                Err(Error::last_os_error())
```

Reply:
The [manual](https://man7.org/linux/man-pages/man7/bpf-helpers.7.html) reads
> Return 0 on success, or a negative error in case of failure.

And the [code](https://elixir.bootlin.com/linux/v4.5/source/kernel/trace/bpf_trace.c#L226) is quite clear: errno isn't involved here, it does things like
```c
if (unlikely(event->oncpu != smp_processor_id()))
		return -EOPNOTSUPP;
```

Reply:
That's annoying, well I suppose you can use [from_raw_os_error](https://doc.rust-lang.org/std/io/struct.Error.html#method.from_raw_os_error) with the value directly then...